### PR TITLE
Restore IP forwarding for Sponge 8

### PR DESCRIPTION
**Pull request by astei**
A few notable changes:

* Velocity modern forwarding support has been added (in fact, it's the recommended solution)
* Simplified the configuration a bit - if the user enables the BungeeCord module, we assume they do want forwarding

I'm creating this as a draft as this is a bit of functionality that requires configuration to work, but configuration is currently busted.

Thanks to @Zidane for answering my n00b questions about Mixins, this is the first time I've ever made a mixin. I unfortunately needed to use an access transformer to accomplish the Velocity modern forwarding support. I talked to Zidane and he said this was my only option currently.

TODO:
- [ ] Make sure configuration works
- [ ] Move to just an ipforwarding module and reorganize the configuration along those lines
- [ ] Evaluate how #3065 impacts this

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -3,3 +3,4 @@ distributionPath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+org.gradle.jvmargs=-Xms512M -Xmx4g -XX:MaxPermSize=1024m -XX:MaxMetaspaceSize=1g -Dkotlin.daemon.jvm.options="-Xmx1g"
```
Comment (Zidane) [inline gradle/wrapper/gradle-wrapper.properties]: Did you mean to commit this?

Comment (astei) [reply gradle/wrapper/gradle-wrapper.properties]: Nope, I've reverted this.

File: src/mixins/java/org/spongepowered/common/mixin/ipforward/velocity/network/ServerLoginNetHandlerMixin_Velocity.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            NetworkManagerAccessor networkManagerAccessor = (NetworkManagerAccessor) this.networkManager;
+            networkManagerAccessor.accessor$setSocketAddress(new InetSocketAddress(VelocityForwardingInfo.readAddress(buf), ((InetSocketAddress) this.networkManager.getRemoteAddress()).getPort()));
+
+            this.loginGameProfile = VelocityForwardingInfo.createProfile(buf);
+            this.currentLoginState = ServerLoginNetHandler.State.READY_TO_ACCEPT;
+            info.cancel();
+        }
+    }
+
+    @Shadow public abstract void disconnect(ITextComponent reason);
```
Comment (Zidane) [inline src/mixins/java/org/spongepowered/common/mixin/ipforward/velocity/network/ServerLoginNetHandlerMixin_Velocity.java]: Shadows go at the top, thinking the plugin put it there on ya.

Comment (Zidane) [reply src/mixins/java/org/spongepowered/common/mixin/ipforward/velocity/network/ServerLoginNetHandlerMixin_Velocity.java]: Also prefix shadow methods with shadow$ please.

Comment (astei) [reply src/mixins/java/org/spongepowered/common/mixin/ipforward/velocity/network/ServerLoginNetHandlerMixin_Velocity.java]: The plugin cursed me!

**Comment by Cybermaxke** [comment]
This will conflict with changes I made in https://github.com/SpongePowered/SpongeCommon/pull/3065, the velocity channel will need to hook into this system instead once merged.

**Comment by astei** [comment]
Looks like I'll need to pick this up again.

**Comment by zml2008** [comment]
Will you be able to finish this up soon? If not, I don't mind adding this to my list to get done.

**Comment by astei** [comment]
Unfortunately, no. I'm busy with other stuff now (new job for instance). Feel free to pick up where I left off.

**Comment by zml2008** [comment]
No worries, thanks for the quick response :)

**Comment by zml2008** [comment]
Replaced by #3302, thanks for your work getting this going!

**Pull request closed without merging**